test: make assert_return() critical by default

yuwata opened the pull request:
<!-- devel-freezer = {"comment-id":"1813698627","freezing-tag":"v255-rc4"} -->

evverx commented on src/basic/log.c:
```diff
@@ -1310,6 +1339,10 @@ void log_parse_environment_variables(void) {
         if (e && log_set_ratelimit_kmsg_from_string(e) < 0)
                 log_warning("Failed to parse log ratelimit kmsg boolean '%s'. Ignoring.", e);
+
+        e = getenv("SYSTEMD_LOG_ASSERT_RETURN_IS_CRITICAL");
```
I feel a bit uneasy about this switch given that it can start crashing systemd left and right :-) Another option would be a compile-time option that could be passed by the CI. (at least that's what NetworkManager did as far as I can remember)

evverx replied:
Looks like NM flipped it at runtime to judge from https://github.com/systemd/systemd/pull/22406#issuecomment-1029391091. I still think it should be a compile-time switch though.

mrc0mmand replied:
Yeah, I was planning to introduce a build time option and explicitly enable it in the CI, since this has a very high potential of screwing up the whole system.

evverx replied:
I think it would actually be nice if it was possible to control it at runtime too because I can imagine scenarios where it would be useful to crash resolved (or any other program linked against libsystemd) without touching networkd but that runtime switch should be hidden behind a compile time switch probably :-)

evverx replied:
Looks like `BUILD_MODE_DEVELOPER` should suffice here. It's already used to turn on additional assertions and it's on by default on the CI (except for Ubuntu CI and Packit).

yuwata replied:
The environment variable and kernel option are dropped. Now, the flag is unconditionally enabled in tests and fuzzers, and enabled in all components when build with `-Dmode=developer` (default).

evverx replied:
I think it would make sense to keep the environment variable and turn it off by default to make it easier to test separate things separately. If I tested a hypothetical NetworkManager I probably wouldn't want `systemd` to go down along with VMs/containers. The systemd CI can always turn it on everywhere. Then again it can't always control external code linked against libsystemd so it could be useful to turn it off selectively.

yuwata replied:
Why not to use release mode build when you want to test other software, e.g. NetworkManager?

evverx replied:
Because `assert_return` isn't supposed to crash in release mode but I'd want it to crash on the CI. For example here's what NM did according to https://github.com/systemd/systemd/pull/22406#issuecomment-1029403606
> NM redefines assert_return() to behave like g_return_if_fail()... which under normal conditions only prints a message.
But in NM-CI we run with G_DEBUG=fatal-warnings, which turns this into an abort().

(NM replaced those systemd parts so it's a hypothetical example)

evverx replied:
I took a closer look at NM and it seems the dhcpv6 client is still there so maybe it isn't a hypothetical example after all :-) Either way systemd crashes aren't that unusual to me so I don't feel strongly about this. It's just that I'm not sure what is going to happen with this option enabled everywhere if some external code linked against libsystemd starts crashing on CentOS CI once it pulls the latest version of Arch or something like that. I'll leave it to @mrc0mmand to decide :-)

evverx replied:
> I don't understand. We have hundreds of command-line switches which will cause the system to fail if used incorrectly

It's kind of hard to set them all to get almost all the systemd components to crash at once.

> not as bad as many others

Dunno. I turned this on globally at some point and rolled it back because even in the environment prepared to deal with all sorts of systemd crashes where PID1, journald, resolved, networkd and so on go down on a regular basis it was unbearable. I guess if weird packets are never sent or other weird things like spawning a gazillion of weird jobs are done it can be somewhat tolerated probably. Either way I think it's a debug feature and it should be available in debug builds only by analogy with the other additional assertions.

evverx replied:
That being said it it was in release mode it would make it easier to set it outside of, say, various CIs but I'm not sure who would need it because to, for example, instrument systemd it would still be necessary to compile systemd separately. All in all I'm for the runtime option available in debug builds only but I don't feel strongly about this.

keszybz reviewed (commented):
Oh, I had some pending comments from a long time ago. I'll publish them, maybe they'll still be useful.

yuwata commented:
Let's add env variable and command line option when necessary. PTAL.

evverx commented:
> Let's add env variable and command line option when necessary

Makes sense.

I'm not sure why the CI is green though :-) With this patch applied `resolved` crashes with
<details>
  <summary>Assertion 'e->state != SD_EVENT_FINISHED' failed at src/libsystemd/sd-event/sd-event.c:1399</summary>

```sh
Program received signal SIGABRT, Aborted.
0x00007f01320b0884 in __pthread_kill_implementation () from /lib64/libc.so.6
(gdb) bt
#0  0x00007f01320b0884 in __pthread_kill_implementation ()
   from /lib64/libc.so.6
#1  0x00007f013205fafe in raise () from /lib64/libc.so.6
#2  0x00007f013204887f in abort () from /lib64/libc.so.6
#3  0x00007f01338d02d6 in log_assert_failed (
    text=0x7f01340009e0 "e->state != SD_EVENT_FINISHED", 
    file=0x7f0133fff403 "src/libsystemd/sd-event/sd-event.c", line=1399, 
    func=0x7f01340045a0 <__func__.148> "sd_event_add_time")
    at ../src/basic/log.c:948
#4  0x00007f01338d0457 in log_assert_failed_return (
    text=0x7f01340009e0 "e->state != SD_EVENT_FINISHED", 
    file=0x7f0133fff403 "src/libsystemd/sd-event/sd-event.c", line=1399, 
    func=0x7f01340045a0 <__func__.148> "sd_event_add_time")
    at ../src/basic/log.c:967
#5  0x00007f0133c7ed83 in sd_event_add_time (e=0x617000022280, 
    ret=0x610000007e98, clock=7, usec=24054941030, accuracy=0, 
    callback=0x4625b4 <on_announcement_timeout>, userdata=0x610000007e40)
    at ../src/libsystemd/sd-event/sd-event.c:1399
#6  0x00007f0133c7f725 in sd_event_add_time_relative (e=0x617000022280, 
    ret=0x610000007e98, clock=7, usec=1000000, accuracy=0, 
    callback=0x4625b4 <on_announcement_timeout>, userdata=0x610000007e40)
    at ../src/libsystemd/sd-event/sd-event.c:1462
#7  0x0000000000464cac in dns_scope_announce (scope=0x610000007e40, goodbye=true) at ../src/resolve/resolved-dns-scope.c:1530
#8  0x0000000000504d08 in link_free (l=0x612000023d40) at ../src/resolve/resolved-link.c:83
#9  0x000000000052dbbd in manager_free (m=0x619000000a80) at ../src/resolve/resolved-manager.c:697
#10 0x0000000000562328 in manager_freep (p=0x7f012f800040) at ../src/resolve/resolved-manager.h:198
#11 0x000000000056315a in run (argc=1, argv=0x7fff22b06468) at ../src/resolve/resolved.c:25
#12 0x0000000000563284 in main (argc=1, argv=0x7fff22b06468) at ../src/resolve/resolved.c:99
```
</details>

and `networkd` crashes with
<details>
  <summary>Assertion 'e->state != SD_EVENT_FINISHED' failed at src/libsystemd/sd-event/sd-event.c:1252</summary>

```sh
#0  0x00007f8b1f6b0884 in __pthread_kill_implementation () from target:/lib64/libc.so.6
#1  0x00007f8b1f65fafe in raise () from target:/lib64/libc.so.6
#2  0x00007f8b1f64887f in abort () from target:/lib64/libc.so.6
#3  0x00007f8b208d02d6 in log_assert_failed (text=0x7f8b210009e0 "e->state != SD_EVENT_FINISHED", file=0x7f8b20fff403 "src/libsystemd/sd-event/sd-event.c", 
    line=1252, func=0x7f8b21004400 <__func__.154> "sd_event_add_io") at ../src/basic/log.c:948
#4  0x00007f8b208d0457 in log_assert_failed_return (text=0x7f8b210009e0 "e->state != SD_EVENT_FINISHED", 
    file=0x7f8b20fff403 "src/libsystemd/sd-event/sd-event.c", line=1252, func=0x7f8b21004400 <__func__.154> "sd_event_add_io") at ../src/basic/log.c:967
#5  0x00007f8b20c7d102 in sd_event_add_io (e=0x617000000080, ret=0x60c000000a20, fd=11, events=1, callback=0x7dfd85 <ipv4acd_on_packet>, 
    userdata=0x60c000000a00) at ../src/libsystemd/sd-event/sd-event.c:1252
#6  0x00000000007e3934 in sd_ipv4acd_start (acd=0x60c000000a00, reset_conflicts=true) at ../src/libsystemd-network/sd-ipv4acd.c:597
#7  0x00000000007e72b9 in ipv4ll_start_internal (ll=0x6080000006a0, reset_generation=true) at ../src/libsystemd-network/sd-ipv4ll.c:278
#8  0x00000000007e7462 in sd_ipv4ll_start (ll=0x6080000006a0) at ../src/libsystemd-network/sd-ipv4ll.c:298
#9  0x00000000006047a1 in dhcp4_handler (client=0x617000000400, event=0, userdata=0x61a000000680) at ../src/network/networkd-dhcp4.c:1183
#10 0x000000000075b1ed in client_notify (client=0x617000000400, event=0) at ../src/libsystemd-network/sd-dhcp-client.c:783
#11 0x000000000075bf8d in client_stop (client=0x617000000400, error=0) at ../src/libsystemd-network/sd-dhcp-client.c:821
#12 0x000000000077710f in sd_dhcp_client_stop (client=0x617000000400) at ../src/libsystemd-network/sd-dhcp-client.c:2388
#13 0x000000000065cdd1 in link_stop_engines (link=0x61a000000680, may_keep_dhcp=true) at ../src/network/networkd-link.c:336
#14 0x000000000041f214 in manager_free (m=0x618000000080) at ../src/network/networkd-manager.c:613
#15 0x00000000004124e3 in manager_freep (p=0x7f8b1c800040) at ../src/network/networkd-manager.h:128
#16 0x00000000004139f6 in run (argc=1, argv=0x7ffffe4522e8) at ../src/network/networkd.c:24
#17 0x0000000000413b20 in main (argc=1, argv=0x7ffffe4522e8) at ../src/network/networkd.c:119
```
</details>

without being exposed to anything unusual.

evverx commented:
FWIW I think the networkd crash has something to do with release messages not being sent by the dhcp client when networkd is stopped but I might be wrong.

yuwata commented:
Fixed the crashes in resolved and networkd on exit, and added checks for similar crashes. Let's see if it works.

evverx commented on src/network/networkd-dhcp4.c:
```diff
@@ -1181,7 +1181,7 @@ static int dhcp4_handler(sd_dhcp_client *client, int event, void *userdata) {
 
                                 r = sd_ipv4ll_start(link->ipv4ll);
-                                if (r < 0)
+                                if (r < 0 && r != -ESTALE) /* On exit, we cannot and should not start sd-ipv4ll. */
```
As far as I can see with this patch applied the dhcp client sends release messages when networkd is stopped:
```
    1   0.000000      0.0.0.0 → 255.255.255.255 DHCP 330 DHCP Discover - Transaction ID 0xe3d84750
    2   0.000327 192.168.125.193 → 192.168.125.203 DHCP 327 DHCP ACK      - Transaction ID 0xe3d84750
    3   5.890927 192.168.125.203 → 192.168.125.193 DHCP 307 DHCP Release  - Transaction ID 0xe3d84750
```
The crash is gone too.

evverx commented:
resolved no longer crashes but I'm not sure what it tries to announce there. If it sends goodbye packets (or something like that) there it should probably send them instead of bailing out. I'd need to take a closer look.

(I think it would be better to move the resolved and networkd patches to separate PRs)

evverx commented:
> added checks for similar crashes

I think it's useful but at least CenOS CI should catch crashes and extract backtraces automatically. I'm not sure why it didn't happen here. The only explanation I can come up with is that it doesn't compile systemd in developer mode but judging by https://github.com/search?q=repo%3Asystemd%2Fsystemd-centos-ci%20meson&type=code `mode=release` isn't passed explicitly anywhere either.

mrc0mmand commented:
> > Let's add env variable and command line option when necessary
> 
> Makes sense.
> 
> I'm not sure why the CI is green though :-) With this patch applied `resolved` crashes with
> Assertion 'e->state != SD_EVENT_FINISHED' failed at src/libsystemd/sd-event/sd-event.c:1399

Turns out this code path is not executed by CI:

```
[260122.822302] testsuite-75.sh[50]: + systemctl stop systemd-resolved.service
[260122.831718] systemd-resolved[1155]: link_free: about to call dns_scope_announce(0x0)
[260122.831805] systemd-resolved[1155]: link_free: about to call dns_scope_announce(0x0)
[260122.831881] systemd-resolved[1155]: link_free: about to call dns_scope_announce(0x0)
[260122.831953] systemd-resolved[1155]: link_free: about to call dns_scope_announce(0x0)
...
```

i.e. we bail out early in `dns_scope_announce()`, thus not triggering the assertion. That's because even though mDNS is turned on by default resolved, the test interface (`dns0`) doesn't have `Multicast=yes`.

Just for completeness, we need mDNS here:
https://github.com/systemd/systemd/blob/bf8726d1ee33047b138f677fe4c72ca9989680e8/src/resolve/resolved-dns-scope.c#L1428-L1429

But we also need a non-NULL scope:
https://github.com/systemd/systemd/blob/bf8726d1ee33047b138f677fe4c72ca9989680e8/src/resolve/resolved-dns-scope.c#L1425-L1426

Which is allocated here:
https://github.com/systemd/systemd/blob/bf8726d1ee33047b138f677fe4c72ca9989680e8/src/resolve/resolved-link.c#L143-L148

Interestingly enough, I can't convince resolved to go through the "announcement phase" during shutdown even with an interface that has both `MulticastDNS=yes` and `Multicast=yes`, but that's probably my lack of knowledge in this area :)

https://github.com/systemd/systemd/blob/bf8726d1ee33047b138f677fe4c72ca9989680e8/src/resolve/resolved-dns-scope.c#L1533-L1545

evverx commented:
> Turns out this code path is not executed by CI

Got it. It can explain why it didn't crash. To hit that assertion resolved had to announce at least one service and "talk" to other mDNS responders running on other machines (announcing their own services and browsing the services announced by resolved).

I'm still not sure why networkd didn't crash though :-)

mrc0mmand commented:
> > Turns out this code path is not executed by CI
> 
> Got it. It can explain why it didn't crash. To hit that assertion resolved had to announce at least one service and "talk" to other mDNS responders running on other machines (announcing their own services and browsing the services announced by resolved).
> 
> I'm still not sure why networkd didn't crash though :-)

I did some very scientific debugging, involving a lot of printfs, and I can't seem to reproduce the crash - but I can get _very_ close. With this network file:

```ini
[Match]
Name=foo99

[Network]
DHCP=ipv4
LinkLocalAddressing=ipv4
KeepConfiguration=no
```

I can get all the way to 

https://github.com/systemd/systemd/blob/1fdab6c3069ab945259d70e22a29e80da8370288/src/libsystemd-network/sd-ipv4ll.c#L292-L299

when stopping networkd, but here the reproducer bails out, since here the link local stuff is already running. I suspect this has to do with following paragraph from `systemd.network(1)` for `LinkLocalAddressing=`:

> An IPv4 link-local address is configured when yes or ipv4 and when DHCPv4 autoconfiguration has been unsuccessful for some time. (IPv4 link-local address autoconfiguration will usually happen in parallel with repeated attempts to acquire a DHCPv4 lease).

That is, we'd need to stop networkd while the link local address is not yet configured, but while also trying to acquire a DHCP lease. But maybe I'm just making stuff up and the truth is somewhere else (up until now I didn't even know IPv4 has link local addresses, I thought that's an IPv6 thing, oh well :)

evverx commented:
I think that crash came from an nspawn container with https://github.com/systemd/systemd/blob/main/network/80-container-host0.network. `networkd` on the host used https://github.com/systemd/systemd/blob/main/network/80-container-ve.network. (As far as I can remember there is a PR somewhere that should start exercising the default configs coming with nspawn).

Either way if `mode=release` isn't passed on CentOS CI I think it should be fine.

yuwata commented:
Could anyone review this, and hopefully set the green label? There may be several rooms to improve our test suites more, but at least this should be a good starting point.

cc @mrc0mmand @evverx @keszybz

mrc0mmand commented:
> Could anyone review this, and hopefully set the green label? There may be several rooms to improve our test suites more, but at least this should be a good starting point.
> 
> cc @mrc0mmand @evverx @keszybz

I'm a bit confused about the additional macros - would you mind extending the commit message(s) with a rationale behind them and when they should be used? I think I understand *what* the macros do, but not completely sure *why* we need them (compared to just making all `assert_return()` calls fail when `-Dmode=developer` is used).

yuwata commented:
Extended commit messages, moved the macros to src/shared/tests.h, and moved them from the first commit to the second commit. Total code change is the same as the previous version.

mrc0mmand commented:
> Extended commit messages, moved the macros to src/shared/tests.h, and moved them from the first commit to the second commit. Total code change is the same as the previous version.

Thank you!
